cleanup and enhancement of JEDEC SPI NOR support

pabigot opened the pull request:
Fixes bad unit addresses in the bindings used to test littlefs.

Changes `jedec,spi-nor` binding to require the `size` property, which is in fact required for the driver to build.  Adds the property to all declared compatible nodes in-tree from which it was missing.

Extends `jedec,spi-nor` binding to identify support for deep-power-down mode, a common feature that can reduce standby mode current consumption by as much as 99.9% (MX25L32 from 10 uA to 3 uA; MX24R64 from 9 uA to 7 nA).

Updates driver to place chip into DPD when it is not active.

pabigot committed:
tests/subsys/fs/littlefs: correct flash partition unit address

Unit addresses for partitions should correspond to the start address.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
dts: jedec,spi-nor: require size property

The SPI NOR driver requires that the size (in bits) be provided in the
devicetree node.  Update the binding to make the property required,
and update all nodes based on the memory chip identified.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: spi_nor: generalize support for ULBPR instruction

Replace the hard-coded check for JEDEC IDs with a property that must
be set in the devicetree entry.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
dts: jedec,spi-nor: add support for deep-power-down specification

Provide information required to allow the driver to put the flash chip
into a deep power down mode.  This can reduce standby current by as
much as 90%.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot committed:
drivers: spi_nor: support deep-power-down mode

Add internal API to enter and exit deep power-down mode.  Add Kconfig
option to return to DPD whenever device is not active.

When device power management becomes more mature it should be possible
to implement it, which would allow use of DPD without having to enter
and exit DPD between consecutive transactions.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

ulfalizer commented on dts/bindings/mtd/jedec,spi-nor.yaml:
```diff
@@ -19,11 +19,71 @@ properties:
+  requires-ulbpr:
+    type: boolean
+    required: false
+    description: Indicates the device requires the ULBPR (0x98) command
```
Could maybe use one of the multi-line syntaxes in https://yaml-multiline.info/ (just discovered that site... pretty specialized) and put a period at the end of the first lines.

ulfalizer replied:
Tell me if the output for these long descriptions gets ugly btw. Could improve it in that case.

pabigot replied:
I'd looked at that site, but it didn't seem to add anything beyond what's normally used.  I think I've used `>` in previous bindings; `>-` doesn't add anything unless the value is actually used in generated documentation somewhere.  Is it?

The vast majority of description text in `dts/bindings` does not have a terminating period.

ulfalizer replied:
Could skip the period if it's meant to be some kind of title though I guess...

ulfalizer replied:
> I'd looked at that site, but it didn't seem to add anything beyond what's normally used.

Nice to have the first line line up with the other lines I think.

> I think I've used `>` in previous bindings; `>-` doesn't add anything unless the value is actually used in generated documentation somewhere. Is it?

Should end up in the generarated header/configuration file. Think it's just `rstrip()`ed at the moment.

ulfalizer replied:
Without looking, I suspect that it gets a bit ugly.

I'll make a PR tomorrow so that it gets output as

```c
/*
 * Long description
 * yada yada
 * yada
 */
```

for multi-line descriptions and as

```c
/* yada yada */
```

for single-line descriptions.

pabigot replied:
I think all these yaml issues have been addressed.

vanwinkeljan commented on drivers/flash/spi_nor.c:
```diff
@@ -166,10 +227,14 @@ static int spi_nor_read(struct device *dev, off_t addr, void *dest,
 
 	SYNC_LOCK();
 
+	(void)exit_dpd(dev);
```
Better implement  device power management for the driver instead of putting the device in low power mode for each transaction and leave the decision to put the device in low power mode to the application code.

pabigot replied:
I'd love to, but it's useless.  There's no way to propagate the dependency information down to the driver unless you're interacting with the flash directly.

pabigot replied:
@vanwinkeljan Please re-check this.

I've added a Kconfig option to allow control of whether deep power-down is entered between transactions.  This should preserve existing behavior, while still allowing my use case of dropping current consumption in applications which rarely use the flash.

If you absolutely insist I will go ahead and implement the device power management (eventually), but as I note in the commit message and above it's not every useful until we get device dependency in place so that clients of the device can ensure it's adequately woken before it's used.  As it stands many applications won't even know (or care) whether their file system is on a SPI NOR device, or how to access that device to control its power consumption.

vanwinkeljan replied:
Thx for the update, adding the Kconfig is fine for me.

vanwinkeljan commented on drivers/flash/spi_nor.c:
```diff
@@ -314,6 +391,7 @@ static int spi_nor_configure(struct device *dev)
 	const struct spi_nor_config *params = dev->config->config_info;
 
 	data->spi = device_get_binding(DT_INST_0_JEDEC_SPI_NOR_BUS_NAME);
+	printk("Dev %s at %p\n", DT_INST_0_JEDEC_SPI_NOR_BUS_NAME, data->spi);
```
Use logging infrastructure instead of printk.

pabigot replied:
Left-over debug message, will remove.

pabigot committed:
drivers: spi_nor: improve documentation and validation on flash layout size

The smallest region that can be erashed is one sector, so setting
SPI_NOR_FLASH_LAYOUT_PAGE_SIZE=2048 will fail at runtime when the
flash page API is used to erase a single (or misaligned) page.  Add a
compile-time check that the requested layout page size is erasable.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

pabigot commented:
Trivial style and comment updates, and addition of a check that the configured flash page layout size is a value that can be erased.  Lack of such a check may have contributed to #19438.

```
drivers: spi_nor: improve documentation and validation on flash layout size

The smallest region that can be erashed is one sector, so setting
SPI_NOR_FLASH_LAYOUT_PAGE_SIZE=2048 will fail at runtime when the
flash page API is used to erase a single (or misaligned) page.  Add a
compile-time check that the requested layout page size is erasable.
```

nvlsianpu commented on boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts:
```diff
@@ -53,6 +53,7 @@ arduino_serial: &uart4 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 0x1000000>;
 	at25sf128a: at25sf128a@0 {
 		compatible = "adesto,at25sf128a", "jedec,spi-nor";
+		size = <134217728>;
```
Isn't it possible to put number in hexadecimal representation here? I believe it will by natural.

pabigot replied:
It really isn't, though that's what I used to do.  The datasheets tend to provide them in decimal, and for normal humans it's a lot easier to understand that `134217728` is 128 Mbit and `67108864` is 64 Mbit than to try to understand what `0x8000000` and `0x400000` mean.

Also the majority of nodes that weren't added by me use decimal, so I decided to follow the common path.

nvlsianpu replied:
I can accept that, better align to majority in this case

nvlsianpu commented on dts/bindings/mtd/jedec,spi-nor.yaml:
```diff
@@ -24,13 +24,68 @@ properties:
+
+      Some devices (Macronix MX25R in particular) wake from deep power
+      down by a timed sequence of CSn toggles rather than the RDPD
+      command.  This property specifies three durations measured in
```
Can you specify here which time is which in DT property sequence? `<x y z>`

pabigot replied:
I thought it was clear, but maybe the update is better.

nvlsianpu reviewed (commented):
will comment more soon

nvlsianpu commented:
Added the `enhancement` label as the patch introduces power-save feature for JEDEC SPI NOR devices.

pabigot commented on drivers/flash/spi_nor.c:
```diff
@@ -110,6 +135,87 @@ static int spi_nor_access(const struct device *const dev,
+		k_sleep(K_MSEC(T_DPDD_MS));
+		/* Assert CSn and wait for tCRDP.
+		 *
+		 * Unfortunately the SPI API doesn't allow us to
```
Yes.  This is certainly shaky, since depending on the delay and the bus frequency the timing requirement might not be met, but without control of CSn directly I don't have an alternative.  (We can't just grab the GPIO and muck with it because some drivers don't provide CSn GPIO information.)  It's also possible but unlikely that sending the command would cause the chip to misbehave.  I could see substituting `RDID` instead.

Experimentally this works on at least two devices that have this (efr32mg_sltb002a's MX25R80 support isn't enabled in tree yet, but I've tested it).

pabigot replied:
Switched to using `RDID`, just in case.

pabigot commented on drivers/flash/spi_nor.c:
```diff
@@ -110,6 +135,87 @@ static int spi_nor_access(const struct device *const dev,
+	if (IS_ENABLED(DT_INST_0_JEDEC_SPI_NOR_HAS_DPD)) {
+#ifdef DT_INST_0_JEDEC_SPI_NOR_DPD_WAKEUP_SEQUENCE
+		/* Delay for tDPDD */
+		k_sleep(K_MSEC(T_DPDD_MS));
```
Yeah, ok, we could do that.  I'd have to capture the uptime clock in `enter_dpd` then do some calculations here.  Off-hand that seems doable.  I'll wait until the complete set of requested changes is in place before tackling it, though.

pabigot replied:
I've added support to capture the time at which DPD was entered, and delay where necessary before exiting.  This also allowed eliminating the unconditional wait for enter DPD to complete before proceeding.

nvlsianpu commented on dts/bindings/mtd/jedec,spi-nor.yaml:
```diff
@@ -24,13 +24,68 @@ properties:
   requires-ulbpr:
```
satisfied

nvlsianpu reviewed (commented):
Looks good now, I will approve once CI become green.

pabigot commented:
> Looks good now, I will approve once CI become green.

Finally green (rebased and pushed to clear the stalled tasks).

nvlsianpu commented:
@ulfalizer Can you revisit?

galak commented on samples/drivers/spi_flash/sample.yaml:
```diff
@@ -16,3 +16,10 @@ tests:
     depends_on: spi
+  sample.driver.spi_flash_dpd:
+    tags: spi flash
+    filter: DT_INST_0_JEDEC_SPI_NOR_LABEL
```
This `filter:` needs updating, we removed support for `DT_` symbol based filters, need to use a function like `filter: dt_compat_enabled("jedec,spi-nor")`

tbursztyka commented on drivers/flash/spi_nor.c:
```diff
@@ -110,6 +185,88 @@ static int spi_nor_access(const struct device *const dev,
+#ifdef CONFIG_MULTITHREADING
+	struct spi_nor_data *const driver_data = dev->driver_data;
+
+	k_sem_take(&driver_data->sem, K_FOREVER);
```
sem attribute seem to be always present, so you could use IS_ENABLED(CONFIG_MULTITHREAD) here

tbursztyka commented on drivers/flash/spi_nor.c:
```diff
@@ -110,6 +185,88 @@ static int spi_nor_access(const struct device *const dev,
+#ifdef CONFIG_MULTITHREADING
+	struct spi_nor_data *const driver_data = dev->driver_data;
+
+	k_sem_give(&driver_data->sem);
```
same here

tbursztyka commented on drivers/flash/spi_nor.c:
```diff
@@ -33,19 +69,58 @@ struct spi_nor_data {
 	struct k_sem sem;
 };
 
-#if defined(CONFIG_MULTITHREADING)
```
actually, this SYNC_* macros being replaced by new functions should be in its own commit. It is not related to deep power down.

pabigot replied:
It sorta is, since it's the added complexity of handling DPD that makes the macros unwieldy.  However, if you really want it that way I can do it.

pabigot committed:
drivers: spi_nor: refactor device mutex management

Convert the LOCK/UNLOCK macros to acquire/release functions in
preparation for extending those steps to include power management.

Also commit to always allocating a semophore, and use a more clean
way of conditionalizing the operations.

Signed-off-by: Peter A. Bigot <pab@pabigot.com>

galak commented on boards/arm/mimxrt1015_evk/mimxrt1015_evk.dts:
```diff
@@ -53,6 +53,7 @@ arduino_serial: &uart4 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 0x1000000>;
 	at25sf128a: at25sf128a@0 {
 		compatible = "adesto,at25sf128a", "jedec,spi-nor";
+		size = <134217728>;
```
its a nit, but for readability can we use DT_SIZE_M

pabigot replied:
Could be confusing, since the size is measured in bits.  This one would have to be `DT_SIZE_M(128 * 8)`.  Also DT_SIZE_M appears to be Zephyr-specific, which isn't ideal.

Do you want to require this?  For consistency we should also change the existing `size` properties in other devicetree files.

galak replied:
Ah, didn't realize size was in bits... (just curious, is it ever not a multiple of 8?).

pabigot replied:
It's never not a multiple of 4096 * 8 since the sector size is always 4096 bytes.  Almost certainly never not a multiple of 65536 * 8, which is the standard block erase size (some chips support 32768 byte block erases).

In early days I used the hex representation, but switched to decimal because it was what most other people were using, and because it's easier for a human to recognize the size.  Few people can look at 0x8000000 and know immediately that it's a 128 Mbit (16 MByte) flash.

pabigot replied:
Also, to look back at history: when I converted to DTS I made the size bits instead of bytes because the datasheets always specify the size in bits rather than bytes.  It's common in some applications to use multiple flash devices in parallel: bits makes more sense in that situation.

galak replied:
Can we do something like:

`size = <134217728>; /* in bits */`

pabigot replied:
Of course we could.  But why?  The property description in the binding specifies that the value is in bits.  A wrong value will be discovered quite rapidly.

This sort of thing is a slippery slope where for consistency we'd end up going through and adding a units value to every numeric device tree property that has units (timeout, other sizes, ...).  I don't really think that's a bad idea, but I also don't think it should be done ad hoc.

galak commented on boards/arm/mimxrt1020_evk/mimxrt1020_evk.dts:
```diff
@@ -61,6 +61,7 @@ arduino_serial: &uart2 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
 	is25wp064: is25wp064@0 {
 		compatible = "issi,is25wp064", "jedec,spi-nor";
+		size = <67108864>;
```
its a nit, but for readability can we use DT_SIZE_M

galak commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk_qspi.dts:
```diff
@@ -12,6 +12,7 @@
 	reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
 	is25wp064: is25wp064@0 {
 		compatible = "issi,is25wp064", "jedec,spi-nor";
+		size = <67108864>;
```
its a nit, but for readability can we use DT_SIZE_M

galak commented on boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts:
```diff
@@ -68,6 +68,7 @@ arduino_serial: &uart3 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 0x800000>;
 	is25wp064: is25wp064@0 {
 		compatible = "issi,is25wp064", "jedec,spi-nor";
+		size = <67108864>;
```
its a nit, but for readability can we use DT_SIZE_M

galak commented on boards/riscv/hifive1/hifive1.dts:
```diff
@@ -63,6 +63,7 @@
 	reg = <0x10014000 0x1000 0x20400000 0xc00000>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
its a nit, but for readability can we use DT_SIZE_M

galak commented on boards/riscv/hifive1_revb/hifive1_revb.dts:
```diff
@@ -61,6 +61,7 @@
 	reg = <0x10014000 0x1000 0x20010000 0x3c0900>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
its a nit, but for readability can we use DT_SIZE_M

galak commented on boards/riscv/qemu_riscv32/qemu_riscv32.dts:
```diff
@@ -39,6 +39,7 @@
 	reg = <0x10014000 0x1000 0x20400000 0xc00000>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
its a nit, but for readability can we use DT_SIZE_M

galak commented on boards/riscv/qemu_riscv64/qemu_riscv64.dts:
```diff
@@ -43,6 +43,7 @@
 	reg = <0x10014000 0x1000 0x20400000 0xc00000>;
 	flash0: flash@0 {
 		compatible = "issi,is25lp128", "jedec,spi-nor";
+		size = <134217728>;
```
its a nit, but for readability can we use DT_SIZE_M

galak reviewed (commented):
Looks good, minor nit about using DT_SIZE_M macro to make things a bit more readable.

tbursztyka commented on drivers/flash/spi_nor.c:
```diff
@@ -348,11 +490,19 @@ static int spi_nor_configure(struct device *dev)
 #endif /* DT_INST_0_JEDEC_SPI_NOR_CS_GPIOS_CONTROLLER */
 
+	/* Might be in DPD if system restarted without power cycle. */
+	(void)exit_dpd(dev);
```
If you add such (void), maybe you'll need it everywhere where relevant?

pabigot replied:
Easier to remove it than try to explain why it makes more sense here than other places, so it's gone.  Thanks for the approval.

tbursztyka replied:
ok, I was also wondering why not just removing it. We do not enforce returned code check everywhere after all

pabigot replied:
OK, so the reason I did it is that here I really don't care if it fails.  We've already confirmed that we can communicate with the device, and we only need to do this to handle a case where a soft reset with DPD active switched to code that doesn't enable DPD when idle.

The draft Zephyr coding standard does require checking return values everywhere (though the estimate of 186 violations in the code base seems ridiculously low) so all this may have to change anyway.  (Depending on what functions are deemed to "return error information".)

tbursztyka reviewed (approved):
fine by me. just a tiny comment
